File: iphone/MakeReady/Pages/Bible/BibleData.swift
//
//  BibleData.swift
//  MakeReady
//
//  Static data for all 66 books of the Bible.
//  Includes names, abbreviations, chapter counts, categories, and category colors.
//

import SwiftUI
import UIKit

// MARK: - Testament

enum Testament {
    case old
    case new
}

// MARK: - Book Category

enum BookCategory: String, CaseIterable {
    case pentateuch       // Gen–Deut
    case historical       // Josh–Esther
    case wisdom           // Job–Song of Solomon
    case majorProphets    // Isa–Dan
    case minorProphets    // Hos–Mal
    case gospelsAndActs   // Matt–Acts
    case paulineEpistles  // Rom–Philemon
    case generalEpistles  // Heb–Jude
    case apocalyptic      // Revelation

    var color: Color {
        switch self {
        case .pentateuch:      return Color(hex: "#424216")
        case .historical:      return Color(hex: "#143F45")
        case .wisdom:          return Color(hex: "#254422")
        case .majorProphets:   return Color(hex: "#143F45")
        case .minorProphets:   return Color(hex: "#563B2D")
        case .gospelsAndActs:  return Color(hex: "#563139")
        case .paulineEpistles: return Color(hex: "#424216")
        case .generalEpistles: return Color(hex: "#563B2D")
        case .apocalyptic:     return Color(hex: "#563139")
        }
    }

    var uiColor: UIColor {
        switch self {
        case .pentateuch:      return UIColor(red: 0x42/255, green: 0x42/255, blue: 0x16/255, alpha: 1)
        case .historical:      return UIColor(red: 0x14/255, green: 0x3F/255, blue: 0x45/255, alpha: 1)
        case .wisdom:          return UIColor(red: 0x25/255, green: 0x44/255, blue: 0x22/255, alpha: 1)
        case .majorProphets:   return UIColor(red: 0x14/255, green: 0x3F/255, blue: 0x45/255, alpha: 1)
        case .minorProphets:   return UIColor(red: 0x56/255, green: 0x3B/255, blue: 0x2D/255, alpha: 1)
        case .gospelsAndActs:  return UIColor(red: 0x56/255, green: 0x31/255, blue: 0x39/255, alpha: 1)
        case .paulineEpistles: return UIColor(red: 0x42/255, green: 0x42/255, blue: 0x16/255, alpha: 1)
        case .generalEpistles: return UIColor(red: 0x56/255, green: 0x3B/255, blue: 0x2D/255, alpha: 1)
        case .apocalyptic:     return UIColor(red: 0x56/255, green: 0x31/255, blue: 0x39/255, alpha: 1)
        }
    }

    var displayName: String {
        switch self {
        case .pentateuch:      return "Books of the Law"
        case .historical:      return "Historical Books"
        case .wisdom:          return "Wisdom & Poetry"
        case .majorProphets:   return "Major Prophets"
        case .minorProphets:   return "Minor Prophets"
        case .gospelsAndActs:  return "Gospels & Acts"
        case .paulineEpistles: return "Pauline Epistles"
        case .generalEpistles: return "General Epistles"
        case .apocalyptic:     return "Apocalyptic"
        }
    }
}

// MARK: - Bible Book

struct BibleBookInfo: Identifiable {
    let id: Int          // Book number 1–66
    let name: String     // Full name
    let abbreviation: String  // Short label for grid
    let chapters: Int
    let category: BookCategory
    let testament: Testament
}

// MARK: - All Books

let bibleBooks: [BibleBookInfo] = [
    // Old Testament — Pentateuch (1–5)
    BibleBookInfo(id: 1,  name: "Genesis",         abbreviation: "Gen",     chapters: 50, category: .pentateuch,    testament: .old),
    BibleBookInfo(id: 2,  name: "Exodus",          abbreviation: "Exod",    chapters: 40, category: .pentateuch,    testament: .old),
    BibleBookInfo(id: 3,  name: "Leviticus",       abbreviation: "Lev",     chapters: 27, category: .pentateuch,    testament: .old),
    BibleBookInfo(id: 4,  name: "Numbers",         abbreviation: "Num",     chapters: 36, category: .pentateuch,    testament: .old),
    BibleBookInfo(id: 5,  name: "Deuteronomy",     abbreviation: "Deut",    chapters: 34, category: .pentateuch,    testament: .old),

    // Old Testament — Historical (6–17)
    BibleBookInfo(id: 6,  name: "Joshua",          abbreviation: "Josh",    chapters: 24, category: .historical,    testament: .old),
    BibleBookInfo(id: 7,  name: "Judges",          abbreviation: "Judg",    chapters: 21, category: .historical,    testament: .old),
    BibleBookInfo(id: 8,  name: "Ruth",            abbreviation: "Ruth",    chapters: 4,  category: .historical,    testament: .old),
    BibleBookInfo(id: 9,  name: "1 Samuel",        abbreviation: "1 Sam",   chapters: 31, category: .historical,    testament: .old),
    BibleBookInfo(id: 10, name: "2 Samuel",        abbreviation: "2 Sam",   chapters: 24, category: .historical,    testament: .old),
    BibleBookInfo(id: 11, name: "1 Kings",         abbreviation: "1 King",  chapters: 22, category: .historical,    testament: .old),
    BibleBookInfo(id: 12, name: "2 Kings",         abbreviation: "2 King",  chapters: 25, category: .historical,    testament: .old),
    BibleBookInfo(id: 13, name: "1 Chronicles",    abbreviation: "1 Chron", chapters: 29, category: .historical,    testament: .old),
    BibleBookInfo(id: 14, name: "2 Chronicles",    abbreviation: "2 Chron", chapters: 36, category: .historical,    testament: .old),
    BibleBookInfo(id: 15, name: "Ezra",            abbreviation: "Ezra",    chapters: 10, category: .historical,    testament: .old),
    BibleBookInfo(id: 16, name: "Nehemiah",        abbreviation: "Neh",     chapters: 13, category: .historical,    testament: .old),
    BibleBookInfo(id: 17, name: "Esther",          abbreviation: "Esther",  chapters: 10, category: .historical,    testament: .old),

    // Old Testament — Wisdom & Poetry (18–22)
    BibleBookInfo(id: 18, name: "Job",             abbreviation: "Job",     chapters: 42, category: .wisdom,        testament: .old),
    BibleBookInfo(id: 19, name: "Psalms",          abbreviation: "Ps",      chapters: 150, category: .wisdom,       testament: .old),
    BibleBookInfo(id: 20, name: "Proverbs",        abbreviation: "Prov",    chapters: 31, category: .wisdom,        testament: .old),
    BibleBookInfo(id: 21, name: "Ecclesiastes",    abbreviation: "Eccles",  chapters: 12, category: .wisdom,        testament: .old),
    BibleBookInfo(id: 22, name: "Song of Solomon", abbreviation: "Song",    chapters: 8,  category: .wisdom,        testament: .old),

    // Old Testament — Major Prophets (23–27)
    BibleBookInfo(id: 23, name: "Isaiah",          abbreviation: "Isa",     chapters: 66, category: .majorProphets, testament: .old),
    BibleBookInfo(id: 24, name: "Jeremiah",        abbreviation: "Jer",     chapters: 52, category: .majorProphets, testament: .old),
    BibleBookInfo(id: 25, name: "Lamentations",    abbreviation: "Lam",     chapters: 5,  category: .majorProphets, testament: .old),
    BibleBookInfo(id: 26, name: "Ezekiel",         abbreviation: "Ezek",    chapters: 48, category: .majorProphets, testament: .old),
    BibleBookInfo(id: 27, name: "Daniel",          abbreviation: "Dan",     chapters: 12, category: .majorProphets, testament: .old),

    // Old Testament — Minor Prophets (28–39)
    BibleBookInfo(id: 28, name: "Hosea",           abbreviation: "Hos",     chapters: 14, category: .minorProphets, testament: .old),
    BibleBookInfo(id: 29, name: "Joel",            abbreviation: "Joel",    chapters: 3,  category: .minorProphets, testament: .old),
    BibleBookInfo(id: 30, name: "Amos",            abbreviation: "Amos",    chapters: 9,  category: .minorProphets, testament: .old),
    BibleBookInfo(id: 31, name: "Obadiah",         abbreviation: "Obad",    chapters: 1,  category: .minorProphets, testament: .old),
    BibleBookInfo(id: 32, name: "Jonah",           abbreviation: "Jonah",   chapters: 4,  category: .minorProphets, testament: .old),
    BibleBookInfo(id: 33, name: "Micah",           abbreviation: "Micah",   chapters: 7,  category: .minorProphets, testament: .old),
    BibleBookInfo(id: 34, name: "Nahum",           abbreviation: "Nah",     chapters: 3,  category: .minorProphets, testament: .old),
    BibleBookInfo(id: 35, name: "Habakkuk",        abbreviation: "Hab",     chapters: 3,  category: .minorProphets, testament: .old),
    BibleBookInfo(id: 36, name: "Zephaniah",       abbreviation: "Zeph",    chapters: 3,  category: .minorProphets, testament: .old),
    BibleBookInfo(id: 37, name: "Haggai",          abbreviation: "Haggai",  chapters: 2,  category: .minorProphets, testament: .old),
    BibleBookInfo(id: 38, name: "Zechariah",       abbreviation: "Zech",    chapters: 14, category: .minorProphets, testament: .old),
    BibleBookInfo(id: 39, name: "Malachi",         abbreviation: "Mal",     chapters: 4,  category: .minorProphets, testament: .old),

    // New Testament — Gospels & Acts (40–44)
    BibleBookInfo(id: 40, name: "Matthew",         abbreviation: "Matt",    chapters: 28, category: .gospelsAndActs,  testament: .new),
    BibleBookInfo(id: 41, name: "Mark",            abbreviation: "Mark",    chapters: 16, category: .gospelsAndActs,  testament: .new),
    BibleBookInfo(id: 42, name: "Luke",            abbreviation: "Luke",    chapters: 24, category: .gospelsAndActs,  testament: .new),
    BibleBookInfo(id: 43, name: "John",            abbreviation: "John",    chapters: 21, category: .gospelsAndActs,  testament: .new),
    BibleBookInfo(id: 44, name: "Acts",            abbreviation: "Acts",    chapters: 28, category: .gospelsAndActs,  testament: .new),

    // New Testament — Pauline Epistles (45–57)
    BibleBookInfo(id: 45, name: "Romans",          abbreviation: "Rom",     chapters: 16, category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 46, name: "1 Corinthians",   abbreviation: "1 Cor",   chapters: 16, category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 47, name: "2 Corinthians",   abbreviation: "2 Cor",   chapters: 13, category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 48, name: "Galatians",       abbreviation: "Gal",     chapters: 6,  category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 49, name: "Ephesians",       abbreviation: "Eph",     chapters: 6,  category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 50, name: "Philippians",     abbreviation: "Phil",    chapters: 4,  category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 51, name: "Colossians",      abbreviation: "Col",     chapters: 4,  category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 52, name: "1 Thessalonians", abbreviation: "1 Thess", chapters: 5,  category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 53, name: "2 Thessalonians", abbreviation: "2 Thess", chapters: 3,  category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 54, name: "1 Timothy",       abbreviation: "1 Tim",   chapters: 6,  category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 55, name: "2 Timothy",       abbreviation: "2 Tim",   chapters: 4,  category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 56, name: "Titus",           abbreviation: "Titus",   chapters: 3,  category: .paulineEpistles, testament: .new),
    BibleBookInfo(id: 57, name: "Philemon",        abbreviation: "Phil",    chapters: 1,  category: .paulineEpistles, testament: .new),

    // New Testament — General Epistles (58–65)
    BibleBookInfo(id: 58, name: "Hebrews",         abbreviation: "Heb",     chapters: 13, category: .generalEpistles, testament: .new),
    BibleBookInfo(id: 59, name: "James",           abbreviation: "James",   chapters: 5,  category: .generalEpistles, testament: .new),
    BibleBookInfo(id: 60, name: "1 Peter",         abbreviation: "1 Pet",   chapters: 5,  category: .generalEpistles, testament: .new),
    BibleBookInfo(id: 61, name: "2 Peter",         abbreviation: "2 Pet",   chapters: 3,  category: .generalEpistles, testament: .new),
    BibleBookInfo(id: 62, name: "1 John",          abbreviation: "1 John",  chapters: 5,  category: .generalEpistles, testament: .new),
    BibleBookInfo(id: 63, name: "2 John",          abbreviation: "2 John",  chapters: 1,  category: .generalEpistles, testament: .new),
    BibleBookInfo(id: 64, name: "3 John",          abbreviation: "3 John",  chapters: 1,  category: .generalEpistles, testament: .new),
    BibleBookInfo(id: 65, name: "Jude",            abbreviation: "Jude",    chapters: 1,  category: .generalEpistles, testament: .new),

    // New Testament — Apocalyptic (66)
    BibleBookInfo(id: 66, name: "Revelation",      abbreviation: "Rev",     chapters: 22, category: .apocalyptic,     testament: .new),
]

// MARK: - Convenience

let oldTestamentBooks = bibleBooks.filter { $0.testament == .old }
let newTestamentBooks = bibleBooks.filter { $0.testament == .new }

File: client/resources/js/utils/bible-data.ts
// Static Bible metadata — web port of iphone/MakeReady/Pages/Bible/
// BibleData.swift (all 66 books: names, grid abbreviations, chapter counts,
// category colors) and BibleVersionMenu.swift (known translations + the
// popular-first sort order for the version dropdown).

export type BibleTestament = 'old' | 'new'

export type BibleBookCategory =
  | 'pentateuch'
  | 'historical'
  | 'wisdom'
  | 'majorProphets'
  | 'minorProphets'
  | 'gospelsAndActs'
  | 'paulineEpistles'
  | 'generalEpistles'
  | 'apocalyptic'

export interface BibleBookInfo {
  id: number // book number 1–66
  name: string
  abbreviation: string
  chapters: number
  category: BibleBookCategory
  testament: BibleTestament
}

/** BookCategory.color hex values (BibleData.swift). */
export const BIBLE_CATEGORY_COLORS: Record<BibleBookCategory, string> = {
  pentateuch: '#424216',
  historical: '#143F45',
  wisdom: '#254422',
  majorProphets: '#143F45',
  minorProphets: '#563B2D',
  gospelsAndActs: '#563139',
  paulineEpistles: '#424216',
  generalEpistles: '#563B2D',
  apocalyptic: '#563139',
}

const B = (
  id: number,
  name: string,
  abbreviation: string,
  chapters: number,
  category: BibleBookCategory,
  testament: BibleTestament,
): BibleBookInfo => ({ id, name, abbreviation, chapters, category, testament })

export const bibleBooks: BibleBookInfo[] = [
  // Old Testament — Pentateuch (1–5)
  B(1, 'Genesis', 'Gen', 50, 'pentateuch', 'old'),
  B(2, 'Exodus', 'Exod', 40, 'pentateuch', 'old'),
  B(3, 'Leviticus', 'Lev', 27, 'pentateuch', 'old'),
  B(4, 'Numbers', 'Num', 36, 'pentateuch', 'old'),
  B(5, 'Deuteronomy', 'Deut', 34, 'pentateuch', 'old'),
  // Historical (6–17)
  B(6, 'Joshua', 'Josh', 24, 'historical', 'old'),
  B(7, 'Judges', 'Judg', 21, 'historical', 'old'),
  B(8, 'Ruth', 'Ruth', 4, 'historical', 'old'),
  B(9, '1 Samuel', '1 Sam', 31, 'historical', 'old'),
  B(10, '2 Samuel', '2 Sam', 24, 'historical', 'old'),
  B(11, '1 Kings', '1 King', 22, 'historical', 'old'),
  B(12, '2 Kings', '2 King', 25, 'historical', 'old'),
  B(13, '1 Chronicles', '1 Chron', 29, 'historical', 'old'),
  B(14, '2 Chronicles', '2 Chron', 36, 'historical', 'old'),
  B(15, 'Ezra', 'Ezra', 10, 'historical', 'old'),
  B(16, 'Nehemiah', 'Neh', 13, 'historical', 'old'),
  B(17, 'Esther', 'Esther', 10, 'historical', 'old'),
  // Wisdom & Poetry (18–22)
  B(18, 'Job', 'Job', 42, 'wisdom', 'old'),
  B(19, 'Psalms', 'Ps', 150, 'wisdom', 'old'),
  B(20, 'Proverbs', 'Prov', 31, 'wisdom', 'old'),
  B(21, 'Ecclesiastes', 'Eccles', 12, 'wisdom', 'old'),
  B(22, 'Song of Solomon', 'Song', 8, 'wisdom', 'old'),
  // Major Prophets (23–27)
  B(23, 'Isaiah', 'Isa', 66, 'majorProphets', 'old'),
  B(24, 'Jeremiah', 'Jer', 52, 'majorProphets', 'old'),
  B(25, 'Lamentations', 'Lam', 5, 'majorProphets', 'old'),
  B(26, 'Ezekiel', 'Ezek', 48, 'majorProphets', 'old'),
  B(27, 'Daniel', 'Dan', 12, 'majorProphets', 'old'),
  // Minor Prophets (28–39)
  B(28, 'Hosea', 'Hos', 14, 'minorProphets', 'old'),
  B(29, 'Joel', 'Joel', 3, 'minorProphets', 'old'),
  B(30, 'Amos', 'Amos', 9, 'minorProphets', 'old'),
  B(31, 'Obadiah', 'Obad', 1, 'minorProphets', 'old'),
  B(32, 'Jonah', 'Jonah', 4, 'minorProphets', 'old'),
  B(33, 'Micah', 'Micah', 7, 'minorProphets', 'old'),
  B(34, 'Nahum', 'Nah', 3, 'minorProphets', 'old'),
  B(35, 'Habakkuk', 'Hab', 3, 'minorProphets', 'old'),
  B(36, 'Zephaniah', 'Zeph', 3, 'minorProphets', 'old'),
  B(37, 'Haggai', 'Haggai', 2, 'minorProphets', 'old'),
  B(38, 'Zechariah', 'Zech', 14, 'minorProphets', 'old'),
  B(39, 'Malachi', 'Mal', 4, 'minorProphets', 'old'),
  // New Testament — Gospels & Acts (40–44)
  B(40, 'Matthew', 'Matt', 28, 'gospelsAndActs', 'new'),
  B(41, 'Mark', 'Mark', 16, 'gospelsAndActs', 'new'),
  B(42, 'Luke', 'Luke', 24, 'gospelsAndActs', 'new'),
  B(43, 'John', 'John', 21, 'gospelsAndActs', 'new'),
  B(44, 'Acts', 'Acts', 28, 'gospelsAndActs', 'new'),
  // Pauline Epistles (45–57)
  B(45, 'Romans', 'Rom', 16, 'paulineEpistles', 'new'),
  B(46, '1 Corinthians', '1 Cor', 16, 'paulineEpistles', 'new'),
  B(47, '2 Corinthians', '2 Cor', 13, 'paulineEpistles', 'new'),
  B(48, 'Galatians', 'Gal', 6, 'paulineEpistles', 'new'),
  B(49, 'Ephesians', 'Eph', 6, 'paulineEpistles', 'new'),
  B(50, 'Philippians', 'Phil', 4, 'paulineEpistles', 'new'),
  B(51, 'Colossians', 'Col', 4, 'paulineEpistles', 'new'),
  B(52, '1 Thessalonians', '1 Thess', 5, 'paulineEpistles', 'new'),
  B(53, '2 Thessalonians', '2 Thess', 3, 'paulineEpistles', 'new'),
  B(54, '1 Timothy', '1 Tim', 6, 'paulineEpistles', 'new'),
  B(55, '2 Timothy', '2 Tim', 4, 'paulineEpistles', 'new'),
  B(56, 'Titus', 'Titus', 3, 'paulineEpistles', 'new'),
  B(57, 'Philemon', 'Phil', 1, 'paulineEpistles', 'new'),
  // General Epistles (58–65)
  B(58, 'Hebrews', 'Heb', 13, 'generalEpistles', 'new'),
  B(59, 'James', 'James', 5, 'generalEpistles', 'new'),
  B(60, '1 Peter', '1 Pet', 5, 'generalEpistles', 'new'),
  B(61, '2 Peter', '2 Pet', 3, 'generalEpistles', 'new'),
  B(62, '1 John', '1 John', 5, 'generalEpistles', 'new'),
  B(63, '2 John', '2 John', 1, 'generalEpistles', 'new'),
  B(64, '3 John', '3 John', 1, 'generalEpistles', 'new'),
  B(65, 'Jude', 'Jude', 1, 'generalEpistles', 'new'),
  // Apocalyptic (66)
  B(66, 'Revelation', 'Rev', 22, 'apocalyptic', 'new'),
]

export const oldTestamentBooks = bibleBooks.filter((b) => b.testament === 'old')
export const newTestamentBooks = bibleBooks.filter((b) => b.testament === 'new')

export function bibleBookByNumber(bookNumber: number): BibleBookInfo | undefined {
  return bibleBooks.find((b) => b.id === bookNumber)
}

export interface BibleVersionInfo {
  id: string // API.Bible id
  code: string // abbreviation shown in the version badge
  name: string
}

/** Cached fallback translations (BibleVersionMenu.swift knownBibleVersions). */
export const knownBibleVersions: BibleVersionInfo[] = [
  { id: 'de4e12af7f28f599-02', code: 'KJV', name: 'King James (Authorised) Version' },
  { id: '01b29f4b342acc35-01', code: 'ASV', name: 'American Standard Version' },
  { id: '9879dbb7cfe39e4d-04', code: 'FBV', name: 'Free Bible Version' },
  { id: '65eec8e0b60e656b-01', code: 'GNV', name: 'Geneva Bible' },
  { id: '55212e3cf4d04c49-01', code: 'GNTD', name: 'Good News Translation (US Version)' },
  { id: 'b32b9d1b64b4ef29-01', code: 'LSV', name: 'Literal Standard Version' },
  { id: 'c315fa9f71d4af3a-01', code: 'WEB', name: 'World English Bible' },
  { id: 'f72b840c855f362c-04', code: 'WEBBE', name: 'World English Bible, British Edition' },
  { id: '7142879509583d59-04', code: 'T4T', name: 'Translation for Translators' },
  { id: 'bba9f40f2062d3ca-01', code: 'BSB', name: 'Berean Standard Bible' },
]

/** Version-dropdown sort: popular codes first, then by name (iOS versionTapped). */
export const POPULAR_TRANSLATION_CODES = ['KJV', 'NIV', 'ESV', 'NASB', 'NLT', 'NKJV', 'CSB', 'BSB', 'WEB', 'ASV']

export function sortVersionsPopularFirst<T extends { code: string; name: string }>(versions: T[]): T[] {
  return [...versions].sort((a, b) => {
    const ai = POPULAR_TRANSLATION_CODES.indexOf(a.code.toUpperCase())
    const bi = POPULAR_TRANSLATION_CODES.indexOf(b.code.toUpperCase())
    const av = ai === -1 ? Number.MAX_SAFE_INTEGER : ai
    const bv = bi === -1 ? Number.MAX_SAFE_INTEGER : bi
    if (av !== bv) return av - bv
    return a.name < b.name ? -1 : a.name > b.name ? 1 : 0
  })
}

/** PassageData.reference (StudyModels.swift) — "Romans 1:1" / "Romans 1:1-5" / "Romans 1:28 - 2:4". */
export function passageReference(p: {
  bookName: string
  chapterStart: number
  chapterEnd?: number | null
  verseStart: number
  verseEnd: number
}): string {
  if (p.chapterEnd != null && p.chapterEnd !== p.chapterStart) {
    return `${p.bookName} ${p.chapterStart}:${p.verseStart} - ${p.chapterEnd}:${p.verseEnd}`
  }
  if (p.verseStart === p.verseEnd) return `${p.bookName} ${p.chapterStart}:${p.verseStart}`
  return `${p.bookName} ${p.chapterStart}:${p.verseStart}-${p.verseEnd}`
}
